non_preemptive_sjf_scheduling.cpp

```diff
@@ -0,0 +1,296 @@
+                 << get<1>(result[i]) << std::setw(17) << left
+                 << get<2>(result[i]) << std::setw(17) << left
+                 << get<3>(result[i]) << std::setw(17) << left
+                 << get<4>(result[i]) << std::setw(17) << left
+                 << get<5>(result[i]) << endl;
+        }
+    }
+};
+
+    /**
+   * @brief Computes the final status of processes after applying non-preemptive SJF scheduling
+   * @tparam S Data type of Process ID
+   * @tparam T Data type of Arrival time
+   * @tparam E Data type of Burst time
+   * @param input A vector of tuples containing Process ID, Arrival time, and Burst time
+   * @returns A vector of tuples containing Process ID, Arrival time, Burst time,
+   *         Completion time, Turnaround time, and Waiting time
+   */
+   
+    template <typename S, typename T, typename E>
+  vector<tuple<S, T, E, double, double, double>> get_final_status(
+      vector<tuple<S, T, E>> input) {
+    
+      // Sort the processes based on Arrival time and then Burst time
+      sort(input.begin(), input.end(), sortcol<S, T, E>);
+
+      // Result vector to hold the final status of each process
+      vector<tuple<S, T, E, double, double, double>> result(input.size());
+      double timeElapsed = 0;
+
+      for (size_t i = 0; i < input.size(); i++) {
+        // Extract Arrival time and Burst time
+          T arrival = get<1>(input[i]);
+          E burst = get<2>(input[i]);
+
+          // If the CPU is idle, move time to the arrival of the next process
+          if (arrival > timeElapsed) {
+              timeElapsed = arrival;
+          }
+
+          // Update timeElapsed by adding the burst time
+          timeElapsed += burst;
+
+          // Calculate Completion time, Turnaround time, and Waiting time
+          double completion = timeElapsed;
+          double turnaround = completion - arrival;
+          double waiting = turnaround - burst;
+
+          // Store the results in the result vector
+          result[i] = make_tuple(get<0>(input[i]), arrival, burst, completion, turnaround, waiting);
+      }
+    
+      return result;
+  }
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    for (int i{}; i < 1000; i++) {
+        srand(time(nullptr));
+        uint32_t n = 1 + rand() % 1000;
+        SJF<uint32_t, uint32_t, uint32_t> readyQueue;
+        vector<tuple<uint32_t, uint32_t, uint32_t>> input(n);
+
+        for (uint32_t i{}; i < n; i++) {
+            get<0>(input[i]) = i;
+            srand(time(nullptr));
+            get<1>(input[i]) = 1 + rand() % 10000;
+            srand(time(nullptr));
+            get<2>(input[i]) = 1 + rand() % 10000;
+        }
+
+        for (uint32_t i{}; i < n; i++) {
+            readyQueue.addProcess(get<0>(input[i]), get<1>(input[i]),
+                                  get<2>(input[i]));
+        }
+        readyQueue.scheduleForSJF();
```

Comment:
you've added an sjf queue but you havent really asserted that any of its values are where they are after operations

Reply:
the values after the operation can be seen in the next line:

readyQueueprintResult();

Reply:
> the values after the operation can be seen in the next line:
> 
> readyQueueprintResult();

printing is not testing please assert those values as the expected values

Reply:
Ok. Could you please tell me from what I understood is correct or not ?

You want me to print both :
        the processes before the operation AND the rearranged process after the operation.

is that right ?